File: src/Templates/DisjointSet.java
package Templates;

public class DisjointSet
{
	static Node v[];
	
	static class Node
	{
		int r, p;

		Node(int pp)
		{
			r = 0;
			p = pp;
		}
	}

	static int find(int x)
	{
		if (v[x].p != x)
			v[x].p = find(v[x].p);
		return v[x].p;
	}

	static void merge(int x, int y)
	{
		x = find(x);
		y = find(y);

		if (x == y)
			return;

		if (v[x].r < v[y].r)
			v[x].p = y;
		else if (v[x].r > v[y].r)
			v[y].p = x;
		else
		{
			v[y].p = x;
			v[x].r++;
		}
	}
}

File: templates/DisjointSet.cpp
struct DisjointSet {
	struct Node {
		int p, sz;
		Node(int p) : p(p), sz(1) {
		}
	};

	int N;
	vector<Node> v;

	DisjointSet(int N) : N(N) {
		for (int i = 0; i <= N; i++) {
			v.pb(Node(i));
		}
	}
	int find(int x) {
		if (v[x].p != x) {
			v[x].p = find(v[x].p);
		}
		return v[x].p;
	}
	void merge(int x, int y) {
		x = find(x);
		y = find(y);

		if (x == y) {
			return;
		}

		if (v[x].sz > v[y].sz) {
			v[x].sz += v[y].sz;
			v[y].p = x;
		} else {
			v[y].sz += v[x].sz;
			v[x].p = y;
		}
	}
};
